File: examples/flutter-app/lib/showcase.dart
/// Cohort simulator — the "Run showcase" engine (mobile).
///
/// Mirrors `examples/svelte-web/src/lib/showcase.ts` so the web and mobile
/// demos produce matching dashboards. Drives the SDK through a synthetic
/// e-commerce cohort — many synthetic users (switched via `setUser`) with
/// realistic drop-off, branching paths and a spread of performance
/// transactions — so the dashboard's Funnels, Journeys and Performance screens
/// have non-trivial data to show.
///
/// This file is deliberately SDK-free: the driver takes an injected
/// [ShowcaseSink], so the planning + emission logic is a pure unit test
/// (`test/showcase_test.dart`). The real Sauron-backed sink lives in `main.dart`.
library;

import 'dart:math';

/// The ordered funnel. Reuses `checkout_completed` (the manual demo button).
const List<String> funnelSteps = <String>[
  'product_viewed',
  'product_added_to_cart',
  'checkout_started',
  'payment_info_entered',
  'checkout_completed',
];

/// Cumulative retention per funnel step — the target drop-off shape.
const List<double> _retention = <double>[1.0, 0.65, 0.43, 0.28, 0.2];

/// Known transaction ops (mirror of the SDK's transaction op set).
const List<String> transactionOps = <String>[
  'navigation',
  'http',
  'resource',
  'screen_load',
  'custom',
];

const int defaultUsers = 120;
const int maxUsers = 500;

/// One timed operation to record via `trackTransaction`.
class SimTransaction {
  const SimTransaction({
    required this.name,
    required this.op,
    required this.duration,
    this.status,
    this.httpMethod,
    this.httpStatus,
    this.url,
  });

  final String name;
  final String op;
  final Duration duration;
  final String? status;
  final String? httpMethod;
  final int? httpStatus;
  final String? url;
}

/// One planned action in a synthetic user's session.
sealed class PlannedAction {
  const PlannedAction();
}

class EventAction extends PlannedAction {
  const EventAction(this.name, [this.properties]);

  final String name;
  final Map<String, Object?>? properties;
}

class TxnAction extends PlannedAction {
  const TxnAction(this.txn);

  final SimTransaction txn;
}

class UserPlan {
  const UserPlan({required this.id, required this.reached, required this.actions});

  final String id;

  /// How many funnel steps this user reached (1..funnelSteps.length).
  final int reached;
  final List<PlannedAction> actions;
}

/// A user identity as the sink sees it.
class SinkUser {
  const SinkUser(this.id, [this.traits]);

  final String id;
  final Map<String, Object?>? traits;
}

/// The SDK surface the driver needs — injected so this file stays testable.
abstract class ShowcaseSink {
  SinkUser? getUser();
  void setUser(SinkUser? user);
  void track(String name, {Map<String, Object?>? properties});
  void trackTransaction(SimTransaction txn);
  Future<void> flush();
}

class ShowcaseProgress {
  const ShowcaseProgress(this.done, this.total, this.events, this.transactions);

  final int done;
  final int total;
  final int events;
  final int transactions;
}

class FunnelCount {
  FunnelCount(this.name, this.count);

  final String name;
  int count;
}

class ShowcaseSummary {
  const ShowcaseSummary(this.users, this.events, this.transactions, this.funnel);

  final int users;
  final int events;
  final int transactions;
  final List<FunnelCount> funnel;
}

/// Deterministic PRNG for a given seed — same seed, same cohort.
Random makeRng(String seed) => Random(seed.hashCode & 0x7fffffff);

/// A right-skewed latency in ms: most values low, a long tail toward [max].
int _skewed(Random rng, int min, int max, [double k = 2.2]) =>
    (min + (max - min) * pow(rng.nextDouble(), k)).round();

/// Plan one synthetic user's actions. Pure — no SDK calls.
UserPlan planUser(Random rng, int index, String runId) {
  int reached = 1;
  for (int i = 1; i < funnelSteps.length; i++) {
    final double cond = _retention[i] / _retention[i - 1];
    if (rng.nextDouble() < cond) {
      reached++;
    } else {
      break;
    }
  }

  final List<PlannedAction> actions = <PlannedAction>[];
  void event(String name, [Map<String, Object?>? properties]) =>
      actions.add(EventAction(name, properties));
  void txn(SimTransaction t) => actions.add(TxnAction(t));

  // Entry: some users arrive via search.
  if (rng.nextDouble() < 0.4) {
    event('search_performed', <String, Object?>{'query': 'wireless headphones'});
  }

  // Landing: route load + product list fetch (+ sometimes a screen/resource load).
  txn(SimTransaction(
    name: 'ProductList',
    op: 'navigation',
    duration: Duration(milliseconds: _skewed(rng, 200, 1400)),
  ));
  txn(SimTransaction(
    name: 'GET /api/products',
    op: 'http',
    duration: Duration(milliseconds: _skewed(rng, 80, 1800)),
    httpMethod: 'GET',
    httpStatus: rng.nextDouble() < 0.03 ? 500 : 200,
    url: '/api/products',
  ));
  txn(SimTransaction(
    name: 'ProductListScreen',
    op: 'screen_load',
    duration: Duration(milliseconds: _skewed(rng, 120, 1600)),
  ));
  if (rng.nextDouble() < 0.5) {
    txn(SimTransaction(
      name: 'assets/catalog.json',
      op: 'resource',
      duration: Duration(milliseconds: _skewed(rng, 40, 600)),
    ));
  }

  // Step 0 — everyone views a product.
  event(funnelSteps[0], <String, Object?>{'product_id': 'sku_${index % 40}'});
  if (rng.nextDouble() < 0.3) event('viewed_recommendations');

  // Step 1 — add to cart.
  if (reached >= 2) {
    event(funnelSteps[1], <String, Object?>{'product_id': 'sku_${index % 40}'});
  }

  // Step 2 — start checkout (+ the checkout API call + screen load).
  if (reached >= 3) {
    event(funnelSteps[2]);
    txn(SimTransaction(
      name: 'CheckoutScreen',
      op: 'screen_load',
      duration: Duration(milliseconds: _skewed(rng, 150, 1800)),
    ));
    txn(SimTransaction(
      name: 'POST /api/checkout',
      op: 'http',
      duration: Duration(milliseconds: _skewed(rng, 150, 2600)),
      httpMethod: 'POST',
      httpStatus: rng.nextDouble() < 0.04 ? 500 : 200,
      url: '/api/checkout',
    ));
    if (rng.nextDouble() < 0.2) {
      event('applied_coupon', <String, Object?>{'code': 'SAVE10'});
    }
  }

  // Step 3 — enter payment.
  if (reached >= 4) event(funnelSteps[3]);

  // Step 4 — order complete.
  if (reached >= 5) {
    final double value = ((19 + rng.nextDouble() * 480) * 100).round() / 100;
    event(funnelSteps[4], <String, Object?>{'cart_value': value, 'currency': 'USD'});
  }

  return UserPlan(id: 'sim_${runId}_$index', reached: reached, actions: actions);
}

int _runCounter = 0;
String _defaultRunId() {
  _runCounter += 1;
  return 'run$_runCounter';
}

/// Drive the [sink] through a cohort of [users] synthetic users. Restores the
/// pre-run identity and flushes on the way out, even if something throws.
Future<ShowcaseSummary> runShowcase(
  ShowcaseSink sink, {
  int users = defaultUsers,
  String? runId,
  int yieldEvery = 12,
  void Function(ShowcaseProgress)? onProgress,
}) async {
  final int total = users.clamp(1, maxUsers);
  final String rid = runId ?? _defaultRunId();
  final int step = yieldEvery < 1 ? 1 : yieldEvery;
  final Random rng = makeRng(rid);

  final SinkUser? previousUser = sink.getUser();
  final List<FunnelCount> funnel =
      <FunnelCount>[for (final String n in funnelSteps) FunnelCount(n, 0)];
  int events = 0;
  int transactions = 0;

  try {
    for (int i = 0; i < total; i++) {
      final UserPlan plan = planUser(rng, i, rid);
      sink.setUser(SinkUser(plan.id, <String, Object?>{'plan': 'sim', 'cohort': rid}));

      for (final PlannedAction action in plan.actions) {
        if (action is EventAction) {
          sink.track(action.name, properties: action.properties);
          events++;
          final int s = funnelSteps.indexOf(action.name);
          if (s >= 0) funnel[s].count++;
        } else if (action is TxnAction) {
          sink.trackTransaction(action.txn);
          transactions++;
        }
      }

      final bool last = i == total - 1;
      if (onProgress != null && (i % step == 0 || last)) {
        onProgress(ShowcaseProgress(i + 1, total, events, transactions));
      }
      if (!last && (i + 1) % step == 0) {
        await Future<void>.delayed(Duration.zero);
      }
    }
  } finally {
    sink.setUser(previousUser);
    await sink.flush();
  }

  return ShowcaseSummary(total, events, transactions, funnel);
}

File: examples/svelte-web/src/lib/showcase.ts
/**
 * Cohort simulator — the "Run showcase" engine.
 *
 * Drives the SDK through a synthetic but realistic e-commerce cohort so the
 * dashboard's **Funnels**, **Journeys** and **Performance** screens light up
 * with non-trivial data. A single demo user clicking buttons is one
 * `distinct_id` → a flat, single-path funnel; a cohort of synthetic users
 * (switched via `setUser`) produces real drop-off and a branching Sankey.
 *
 * This module is deliberately **SDK-free**: the driver takes an injected
 * {@link ShowcaseSink}, so the planning + emission logic is unit-testable under
 * plain Node (see `showcase.test.ts`). The real Sauron-backed sink is built in
 * the UI layer.
 */
import type { TransactionInput } from '@sauron/browser';

/** The ordered funnel. Reuses `checkout_completed` (the manual demo button). */
export const FUNNEL_STEPS: readonly string[] = [
  'product_viewed',
  'product_added_to_cart',
  'checkout_started',
  'payment_info_entered',
  'checkout_completed',
];

/** Cumulative retention per funnel step — the target drop-off shape. */
const RETENTION: readonly number[] = [1.0, 0.65, 0.43, 0.28, 0.2];

/** Known transaction ops (mirror of the SDK's `TransactionOp`). */
export const TRANSACTION_OPS: readonly string[] = [
  'navigation',
  'http',
  'resource',
  'screen_load',
  'custom',
];

export const DEFAULT_USERS = 120;
export const MAX_USERS = 500;

export interface PlannedEvent {
  kind: 'event';
  name: string;
  properties?: Record<string, unknown>;
}
export interface PlannedTxn {
  kind: 'txn';
  input: TransactionInput;
}
export type PlannedAction = PlannedEvent | PlannedTxn;

export interface UserPlan {
  id: string;
  /** How many funnel steps this user reached (1..FUNNEL_STEPS.length). */
  reached: number;
  actions: PlannedAction[];
}

/** A user identity as the sink sees it (subset of the SDK user). */
export interface SinkUser {
  id: string;
  traits?: Record<string, unknown>;
}

/** The SDK surface the driver needs — injected so this module stays testable. */
export interface ShowcaseSink {
  getUser(): SinkUser | null;
  setUser(user: SinkUser | null): void;
  track(name: string, properties?: Record<string, unknown>): void;
  trackTransaction(input: TransactionInput): void;
  flush(): Promise<void>;
}

export interface ShowcaseOptions {
  users?: number;
  runId?: string;
  /** Yield to the event loop (and report progress) every N users. */
  yieldEvery?: number;
}

export interface ShowcaseProgress {
  done: number;
  total: number;
  events: number;
  transactions: number;
}

export interface ShowcaseSummary {
  users: number;
  events: number;
  transactions: number;
  funnel: { name: string; count: number }[];
}

/** Deterministic PRNG (mulberry32 seeded via FNV-1a) — same seed, same cohort. */
export function makeRng(seed: string | number): () => number {
  let h = 2166136261 >>> 0;
  const s = String(seed);
  for (let i = 0; i < s.length; i++) {
    h ^= s.charCodeAt(i);
    h = Math.imul(h, 16777619);
  }
  let a = h >>> 0;
  return function next(): number {
    a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

/** A right-skewed latency in ms: most values low, a long tail toward `max`. */
function skewed(rng: () => number, min: number, max: number, k = 2.2): number {
  return Math.round(min + (max - min) * Math.pow(rng(), k));
}

/** Plan one synthetic user's actions. Pure — no SDK calls. */
export function planUser(rng: () => number, index: number, runId: string): UserPlan {
  // How far down the funnel this user gets, honoring the retention curve.
  let reached = 1;
  for (let i = 1; i < FUNNEL_STEPS.length; i++) {
    const cond = RETENTION[i] / RETENTION[i - 1];
    if (rng() < cond) reached++;
    else break;
  }

  const actions: PlannedAction[] = [];
  const event = (name: string, properties?: Record<string, unknown>) =>
    actions.push({ kind: 'event', name, properties });
  const txn = (input: TransactionInput) => actions.push({ kind: 'txn', input });

  // Entry: some users arrive via search.
  if (rng() < 0.4) event('search_performed', { query: 'wireless headphones' });

  // Landing: route load + product list fetch (+ sometimes a bundle resource).
  txn({ name: 'GET /products', op: 'navigation', durationMs: skewed(rng, 200, 1400) });
  txn({
    name: 'GET /api/products',
    op: 'http',
    durationMs: skewed(rng, 80, 1800),
    httpMethod: 'GET',
    httpStatus: rng() < 0.03 ? 500 : 200,
    status: null,
    url: '/api/products',
  });
  if (rng() < 0.5) {
    txn({ name: 'app.bundle.js', op: 'resource', durationMs: skewed(rng, 40, 600) });
  }

  // Step 0 — everyone views a product.
  event(FUNNEL_STEPS[0], { product_id: `sku_${index % 40}` });
  if (rng() < 0.3) event('viewed_recommendations');

  // Step 1 — add to cart.
  if (reached >= 2) event(FUNNEL_STEPS[1], { product_id: `sku_${index % 40}` });

  // Step 2 — start checkout (+ the checkout API call).
  if (reached >= 3) {
    event(FUNNEL_STEPS[2]);
    txn({
      name: 'POST /api/checkout',
      op: 'http',
      durationMs: skewed(rng, 150, 2600),
      httpMethod: 'POST',
      httpStatus: rng() < 0.04 ? 500 : 200,
      status: null,
      url: '/api/checkout',
    });
    if (rng() < 0.2) event('applied_coupon', { code: 'SAVE10' });
  }

  // Step 3 — enter payment.
  if (reached >= 4) event(FUNNEL_STEPS[3]);

  // Step 4 — order complete.
  if (reached >= 5) {
    const value = Math.round((19 + rng() * 480) * 100) / 100;
    event(FUNNEL_STEPS[4], { cart_value: value, currency: 'USD' });
  }

  return { id: `sim_${runId}_${index}`, reached, actions };
}

function clamp(n: number, lo: number, hi: number): number {
  return Math.max(lo, Math.min(hi, n));
}

let runCounter = 0;
function defaultRunId(): string {
  runCounter += 1;
  return `run${runCounter}`;
}

const tick = () => new Promise<void>((resolve) => setTimeout(resolve, 0));

/**
 * Drive the sink through a cohort of `users` synthetic users. Restores the
 * pre-run identity and flushes on the way out, even if something throws.
 */
export async function runShowcase(
  sink: ShowcaseSink,
  opts: ShowcaseOptions = {},
  onProgress?: (p: ShowcaseProgress) => void,
): Promise<ShowcaseSummary> {
  const total = clamp(Math.floor(opts.users ?? DEFAULT_USERS), 1, MAX_USERS);
  const runId = opts.runId ?? defaultRunId();
  const yieldEvery = Math.max(1, opts.yieldEvery ?? 12);
  const rng = makeRng(runId);

  const previousUser = sink.getUser();
  const funnel = FUNNEL_STEPS.map((name) => ({ name, count: 0 }));
  let events = 0;
  let transactions = 0;

  try {
    for (let i = 0; i < total; i++) {
      const plan = planUser(rng, i, runId);
      sink.setUser({ id: plan.id, traits: { plan: 'sim', cohort: runId } });

      for (const action of plan.actions) {
        if (action.kind === 'event') {
          sink.track(action.name, action.properties);
          events++;
          const step = FUNNEL_STEPS.indexOf(action.name);
          if (step >= 0) funnel[step].count++;
        } else {
          sink.trackTransaction(action.input);
          transactions++;
        }
      }

      const last = i === total - 1;
      if (onProgress && (i % yieldEvery === 0 || last)) {
        onProgress({ done: i + 1, total, events, transactions });
      }
      if (!last && (i + 1) % yieldEvery === 0) await tick();
    }
  } finally {
    sink.setUser(previousUser);
    await sink.flush();
  }

  return { users: total, events, transactions, funnel };
}
